### PR TITLE
String together setup functions

### PR DESCRIPTION
**Pull request by peternga**
I think these functions should actually live on the server-side but I need to set up TS node for our server code.

### CODE REVIEW

**Comment by peternga** [comment]
Still just a draft but wanted to get thoughts on whether this bootstrapping logic makes sense. It's all in the `test` function

File: src/client/scripts/lib/setup.ts [hunk 1]
```diff
@@ -18,3 +24,36 @@ export const shuffle = (arr) => {
...
+  return result;
+};
+
+export const test = () => {
+  // Dummy variable, should get this from actual number of players at start
+  const numPlayers = 2;
+  const playerHands = {};
+  const npcHands = {};
+
+  const chunks = _splitToChunks(shuffle(getFullDeck()), numPlayers);
```
Comment (peternga) [inline src/client/scripts/lib/setup.ts:42]: I imagine we'll pass off parts of the deck to each human player so they can choose a word to send back to the server for their neighbors

File: src/client/scripts/lib/setup.ts [hunk 2]
```diff
@@ -18,3 +24,36 @@ export const shuffle = (arr) => {
...
+  const deck = [];
+
+  // This part should be handled by player interaction
+  // For now, just assign players 5 random letters each
+  for (const index in chunks) {
+    const word = chunks[index].splice(0, 5);
```
Comment (peternga) [inline src/client/scripts/lib/setup.ts:48]: If we allow the players to send the server their neighbor's words, we might need to have some sort of waiting screen after you submit the word but others haven't.

File: src/client/scripts/lib/setup.ts [hunk 3]
```diff
@@ -18,3 +24,36 @@ export const shuffle = (arr) => {
...
+    playerHands[index] = word;
+    deck.push(...chunks[index]);
+  }
+
+  // Take the remaining letters to populate NPC hands
+  for (let i = 0; i < MaxPlayers - numPlayers; i++) {
+    npcHands[i] = deck.splice(0, BaseNPCCards + NPCCardGrowth * i);
```
Comment (peternga) [inline src/client/scripts/lib/setup.ts:55]: Hopefully this is clear. `splice` mutates the array and that's how we have the rest of the deck ready after allocating all the NPC hands.

However, I feel like this is somewhat fragile since it depends on `deck` having an adequate number of cards available and that we aren't generating more NPCs than we can support. I think the game was designed in such a way that this should always be true but it still gives me pause.

File: src/client/scripts/lib/setup.ts [hunk 4]
```diff
@@ -18,3 +24,36 @@ export const shuffle = (arr) => {
...
   return arrCopy;
 };
+
+export const _splitToChunks = (array, parts) => {
+  const result = [];
+  for (let i = parts; i > 0; i--) {
+    result.push(array.splice(0, Math.ceil(array.length / i)));
+  }
+  return result;
+};
```
Comment (nathanleiby) [inline src/client/scripts/lib/setup.ts:34]: optionally, we could use lodash for this sort of stuff: 
- https://lodash.com/docs/#chunk
- https://lodash.com/docs/#shuffle

File: src/client/scripts/lib/setup.ts (hunk 1 above)
Comment (nathanleiby) [reply src/client/scripts/lib/setup.ts:42]: That's how it works in the real game. 

There may be some edge cases to handle, e.g. "a player can't determine a word from the given letters", where I think you give them the leftover cards that other players haven't used in their words?

File: src/client/scripts/lib/setup.ts (hunk 3 above)
Comment (nathanleiby) [reply src/client/scripts/lib/setup.ts:55]: If uncertain about edge cases, perhaps could write some unit tests to validate various cases, e.g. "given X players, expect final state (# of cards per player and NPC) to be ..."

I would suggest we only support the # of players and deck conditions from the base game at first, but perhaps we can add "more players" as a future extension

**Comment by peternga** [comment]
Thanks for the suggestions!
http://g.recordit.co/pMSZvxekcA.gif

I think we'll need to figure out how we want to structure some of these things. It's a bit sloppy right now but it's working.
